Add per-task metrics

nickrobinson251 opened the pull request:
Close https://github.com/JuliaLang/julia/issues/47351 (builds on top of https://github.com/JuliaLang/julia/pull/48416)

Adds two per-task metrics:
- running time = amount of time the task was actually running (according to our scheduler). Note: currently inclusive of GC time, but would be good to be able to separate that out (in a future PR)
-  wall time = amount of time between the scheduler becoming aware of this task and the task entering a terminal state (i.e. done or failed).

We record running time in `wait()`, where the scheduler stops running the task as well as in `yield(t)`, `yieldto(t)` and `throwto(t)`, which bypass the scheduler. Other places where a task stops running (for `Channel`, `ReentrantLock`, `Event`, `Timer` and `Semaphore` are all implemented in terms of `wait(Condition)`, which in turn calls `wait()`. `LibuvStream` similarly calls `wait()`.

This should capture everything (albeit, slightly over-counting task CPU time by including any enqueuing work done before we hit `wait()`).

The various metrics counters could be a separate inlined struct if we think that's a useful abstraction, but for now i've just put them directly in `jl_task_t`. They are all atomic, except the `metrics_enabled` flag itself (which we now have to check on task start/switch/done even if metrics are not enabled) which is set on task construction and marked `const` on the julia side.

In future PRs we could add more per-task metrics, e.g. compilation time, GC time, allocations, potentially a wait-time breakdown (time waiting on locks, channels, in the scheduler run queue, etc.), potentially the number of yields.

Perhaps in future there could be ways to enable this on a per-thread and per-task basis. And potentially in future these same timings could be used by `@time` (e.g. writing this same timing data to a ScopedValue like in https://github.com/JuliaLang/julia/pull/55103 but only for tasks lexically scoped to inside the `@time` block).

Timings are off by default but can be turned on globally via starting Julia with `--task-metrics=yes` or calling `Base.Experimental.task_metrics(true)`. Metrics are collected for all tasks created when metrics are enabled. In other words, enabling/disabling timings via `Base.Experimental.task_metrics` does not affect existing `Task`s, only new `Task`s.

The other new APIs are `Base.Experimental.task_running_time_ns(::Task)` and `Base.Experimental.task_wall_time_ns(::Task)` for retrieving the new metrics. These are safe to call on any task (including the current task, or a task running on another thread). All these are in `Base.Experimental` to give us room to change up the APIs as we add more metrics in future PRs (without worrying about release timelines).

cc @NHDaly @kpamnany @d-netto

vilterp committed:
basic task cpu time tracking

vilterp committed:
initialize to zero

vilterp committed:
make new task field visible from Julia

vilterp committed:
add start time

vilterp committed:
fix typo

vilterp committed:
add test

vilterp committed:
add a wait

vilterp committed:
import `LinearAlgebra` for `peakflops`

vilterp committed:
update last scheduled on task finish as well

vilterp committed:
no need to update last scheduled in finish

vilterp committed:
add test

vilterp committed:
fix test

nickrobinson251 committed:
Add per-task wall-time

nickrobinson251 committed:
Test individual task cpu_time less than wall_time

nickrobinson251 committed:
More tests

nickrobinson251 committed:
Move task timings mostly to julia

nickrobinson251 committed:
Record cpu time on entrance to `wait/yield`

nickrobinson251 committed:
Rename 'scheduled_at' -> '[en/de]queued_at'

nickrobinson251 committed:
Ability to enable/disable task timings globally

nickrobinson251 committed:
fixup whitespace

nickrobinson251 committed:
Enable task timing in test

giordano commented on doc/src/manual/command-line-interface.md:
```diff
@@ -217,6 +217,7 @@ The following is a complete list of command-line switches available when launchi
 |`--trace-compile={stderr\|name}`       |Print precompile statements for methods compiled during execution or save to stderr or a path. Methods that were recompiled are printed in yellow or with a trailing comment if color is not supported|
 |`--trace-compile-timing`               |If --trace-compile is enabled show how long each took to compile in ms|
 |`--trace-dispatch={stderr\|name}`      |Print precompile statements for methods dispatched during execution or save to stderr or a path.|
+|`--task-timing`                        |Enable tasks the collection of per-task timing metrics|
```
Should this be
```suggestion
|`--task-timing`                        |Enable the collection of per-task timing metrics|
```
?

nickrobinson251 commented on src/init.c:
```diff
@@ -874,6 +874,9 @@ static NOINLINE void _finish_julia_init(JL_IMAGE_SEARCH rel, jl_ptls_t ptls, jl_
 
+    if (jl_options.task_timing) {
+        jl_atomic_fetch_add(&jl_task_timing_enabled, 1);
+    }
```
wasn't sure where to was best to do this (i put it in `_finish_julia_init` since that's where most of the user-configurable stuff is, and then plumped for "before `jl_start_threads`"), but actually maybe we should do this before `jl_init_root_task`?

vtjnash reviewed (commented):
SGTM. I believe you may need to add the reporting functions to public and docs, and the PR to NEWs

kpamnany commented on src/jlapi.c:
```diff
@@ -809,6 +809,33 @@ JL_DLLEXPORT uint64_t jl_cumulative_recompile_time_ns(void)
+ */
+JL_DLLEXPORT void jl_task_timing_disable(void)
+{
+    jl_atomic_fetch_add(&jl_task_timing_enabled, -1);
```
```suggestion
    uint8_t tt_enabled = jl_atomic_load_relaxed(&jl_task_timing_enabled);
    if (tt_enabled > 0) {
        jl_atomic_fetch_add(&jl_task_timing_enabled, -1);
    }
    else {
        jl_error("task timing is already disabled");
    }
```

NHDaly replied:
Good call. Though i guess this could still go negative if two different threads try to decrement it at the same time... 🤔
If you want to be really careful you could check-and-set in a loop?
```
    uint8_t tt_enabled = jl_atomic_load_relaxed(&jl_task_timing_enabled);
    while (tt_enabled > 0) {
        jl_atomic_cmpswap(&jl_task_timing_enabled, &tt_enabled, tt_enabled-1);
    }
```
And actually, i don't think erroring sounds nice to me - i wouldn't want to get an error if i try to disable timing and it's already disabled. I think i'd just want it to silently return. 👍

NHDaly replied:
And actually, i don't think erroring sounds nice to me - i wouldn't want to get an error if i try to disable timing and it's already disabled. I think i'd just want it to silently return.

kpamnany commented on src/jlapi.c:
```diff
@@ -809,6 +809,33 @@ JL_DLLEXPORT uint64_t jl_cumulative_recompile_time_ns(void)
+    jl_atomic_fetch_add(&jl_task_timing_enabled, -1);
+}
+
+// TODO: remove this
```
Agree on the TODO comment, but do please remove all TODOs.

kpamnany commented on src/jltypes.c:
```diff
@@ -3760,8 +3760,13 @@ void jl_init_types(void) JL_GC_DISABLED
-                                        "priority"),
-                        jl_svec(16,
+                                        "priority",
+                                        "is_timing_enabled",
+                                        "first_enqueued_at",
+                                        "last_dequeued_at",
+                                        "cpu_time_ns",
+                                        "wall_time_ns"),
```
Should these go in a separate struct called `metrics` or something @vtjnash? We won't be able to initialize that struct type before this, so it would have to be `jl_any_type`, but that would mean we'd have to alloc this in `jl_new_task`. I don't know how/if we can do an embedded struct here?

vtjnash replied:
It could be a separate struct inlined here, but I don't see much value to that. More valuable would probably be marking each of these as atomic though (in jltypes.c) so that observing them isn't UB. Most reads can probably be monotonic, but some writes are probably release and some reads might be acquires

kpamnany replied:
Okay on the separation concern. And yeah, I guess it is necessary for these to be atomic. Will discuss with @nickrobinson251.

nickrobinson251 replied:
Changed to atomics. But i wasn't sure from the docs how to select the correct memory ordering (so you put `:monotonic` everywhere for now) -- would love advice

nickrobinson251 replied:
@vtjnash would you be able to take a look and advise on which orderings to use for the various atomic reads and writes in the PR?

kpamnany commented on src/julia.h:
```diff
@@ -2289,6 +2289,16 @@ typedef struct _jl_task_t {
     // multiqueue priority
     uint16_t priority;
+    // flag indicating whether or not this to measure the wall and cpu time of this task
+    uint8_t is_timing_enabled;
```
```suggestion
    uint8_t timing_enabled;
```
If accepted, there are other places to change that I haven't commented -- will need a search/replace.

kpamnany reviewed (commented):
Nice!

I'm not too satisfied with the metrics being directly in `jl_task_t`; hopefully @vtjnash has a suggestion for what to do there.

We now have to separate out spin time, wait time, compilation time, and GC time (in another PR I guess) and the task CPU time will be actual user CPU time.

IanButterworth commented:
It'd be helpful to add an example of using this to the original post.

nickrobinson251 commented on base/timing.jl:
```diff
@@ -86,6 +86,64 @@ end
+This metric is only updated when the task yields or completes.
+See also [`task_wall_time_ns`](@ref).
+
+Will be `UInt64(0)` if task timings are not enabled.
```
or `missing` or `nothing` are also possible options for that case

NHDaly replied:
IMO you should not use unsigned integers for representing positive values - you should use `Int64` for this instead.

It's better in julia than it is in C, since julia's defaults are less dangerous, but using unsigned ints for positive numeric values is a consistent source of bugs in C programs, and i think should be still frowned upon in julia. (https://jacobegner.blogspot.com/2019/11/unsigned-integers-are-dangerous.html)

IMO Unsigned ints should be used only for things like hashes, identifiers, addresses, and possibly for indexes but even for indexes i'd use signed ints.

But anyway, another benefit is that this also gives `-1` as a possible return value for the unset case.

nickrobinson251 replied:
I've so far opted for returning an `Int` when enabled and `nothing` otherwise. I think `nothing` is maybe nicer because it'll force an error if you try to use it like an `Int`.

That said, these functions currently have` _ns()` names, similar to `cumulative_compile_time_ns()` and `gc_time_ns()` and `time_ns()` which all return `UInt64`, so don't love the inconsistency

I also moved these all to `Experimental` so we can change these APIs as we learn more about what makes sense (as we add more metrics).

nickrobinson251 commented on doc/man/julia.1:
```diff
@@ -294,6 +294,10 @@ If --trace-compile is enabled show how long each took to compile in ms
 
+.TP
+--task-timing
+Enabled the collection of per-task timing metrics.
```
thanks!

thoughts on name, perhaps `--task-metrics` so it makes sense for if/when we add allocation tracking?

kpamnany replied:
I like that, yeah.

NHDaly commented on src/task.c:
```diff
@@ -1146,6 +1146,11 @@ JL_DLLEXPORT jl_task_t *jl_new_task(jl_function_t *start, jl_value_t *completion
     t->ptls = NULL;
     t->world_age = ct->world_age;
     t->reentrant_timing = 0;
+    t->is_timing_enabled = jl_atomic_load_relaxed(&jl_task_timing_enabled) != 0;
```
Oh, interesting, so once a Task is spawned, enabling/disabling the feature doesn't affect live tasks?

So i guess if we wanted to enable this e.g. in `@time`, we would have to explicitly set this field on the Task struct? 🤔 I think that seems okay to me. Either you enable the global field before spawning any Task of interest, or you set the field on the Task when you want to start timing it. 👍 Coolcool.

🤔 And actually, i'm not sure if these numbers would get screwed up if you enable it while a task is running. So maybe you really do need to do this only at Task start?

NHDaly commented on src/julia.h:
```diff
@@ -2289,6 +2289,16 @@ typedef struct _jl_task_t {
     uint16_t priority;
+    // flag indicating whether or not this to measure the wall and cpu time of this task
+    uint8_t is_timing_enabled;
+    // timestamp this task first entered the run queue (TODO: int32 of ms instead?)
+    uint64_t first_enqueued_at;
+    // timestamp this task was most recently scheduled to run
+    uint64_t last_dequeued_at;
```
small nitpicky suggestion, but i am not in love with the name `last_dequeued_at`. I think it is precise, but not as clear as it could be.

Would `last_scheduled_to_run_at` be too verbose? Or `last_scheduled_on_thread_at`? Or `last_started_running_at`?

Also, when picking these names, maybe we can keep in mind that we might add a few more states in the near future for tracking compilation time, and different wait time reasons, so we might end up with things like:
- `last_started_waiting_at`
- `last_started_compiling_at`

etc, so i think i like `last_started_running_at` the best! :)

nickrobinson251 replied:
Happy to change names.

I mostly just want to avoid using the word `scheduled` here, since it can be ambiguous / confusing (since `schedule(t)` means "(re)insert `t` into the scheduler run queue", but when someone asks "when was `t` last scheduled?" they usually mean "when did `t` last start running?" not "when did `t` last get (re)inserted to the run queue?".

`first_enqueued_at` and `last_started_running_at` seem fine to me. 

(These should be internal names anyway, and we should have public APIs like `task_cpu_time()` or whatever that are actually public/stable, but still best if we pick good ones to begin with).

NHDaly replied:
Cool, +1

nickrobinson251 replied:
done

NHDaly commented on src/julia.h:
```diff
@@ -2289,6 +2289,16 @@ typedef struct _jl_task_t {
+    uint64_t first_enqueued_at;
+    // timestamp this task was most recently scheduled to run
+    uint64_t last_dequeued_at;
+    // time this task has spent running; updated when it yields
+    uint64_t cpu_time_ns;
```
```suggestion
    // time this task has spent running; updated when it yields or finishes.
    uint64_t cpu_time_ns;
```

NHDaly commented on test/cmdlineargs.jl:
```diff
@@ -783,6 +783,17 @@ let exename = `$(Base.julia_cmd()) --startup-file=no --color=no`
+    let JL_OPTIONS_TASK_TIMING_OFF = 0, JL_OPTIONS_TASK_TIMING_ON = 1
+        @test parse(Int,readchomp(`$exename -E
+            "Int(Base.JLOptions().task_timing)"`)) == JL_OPTIONS_TASK_TIMING_OFF
+        @test parse(Int, readchomp(`$exename --task-timing -E
+            "Int(Base.JLOptions().task_timing)"`)) == JL_OPTIONS_TASK_TIMING_ON
```
Do you think there's a chance we'll ever want to enable this by default?

And if so, do you think this should be an on/off flag like `--task-timing=on`, rather than `--task-timing`, so that the meaning of the flag can stay unchanged even if we change the defaults in the future? Though i guess we could always introduce a `--no-task-timing` flag in that case in the future.

nickrobinson251 replied:
i wasn't sure whether to prefer `--task-timing` or `--task-timing=yes` but i see the second option leaves us a little more room for flexibility so i'll switch to that

NHDaly commented on base/timing.jl:
```diff
@@ -86,6 +86,64 @@ end
+Task created when Base.task_timing(true) is in effect will have [`task_cpu_time_ns`](@ref)
+and [`task_wall_time_ns`](@ref) timing information available.
+"""
+function task_timing(b::Bool)
+    if b
+        ccall(:jl_task_timing_enable, Cvoid, ())
+    else
+        # TODO: prevent decrementing the counter below zero
+        ccall(:jl_task_timing_disable, Cvoid, ())
+    end
+    return nothing
+end
```
Since you've put effort into making this flag reentrant, maybe you want to provide a `with_task_timing_enabled(f)` variant too? Not sure.

NHDaly commented on test/threads_exec.jl:
```diff
@@ -1312,4 +1313,80 @@ end
+                    cpu_times[$i] = cpu_time_i = Base.task_cpu_time_ns(task_i)
+                    wall_times[$i] = wall_time_i = Base.task_wall_time_ns(task_i)
+                    # task should have recorded some cpu-time and some wall-time
+                    @test cpu_time_i > 0
+                    @test wall_time_i > 0
+                    # task cpu-time cannot be greater than its wall-time
+                    @test wall_time_i >= cpu_time_i
+                    # task wall-time must be less than our manually measured wall-time
+                    # between calling `@spawn` and returning from `wait`.
+                    @test wall_time_delta_i > wall_time_i
```
I don't think you're allowed to put `@test`s in spawned tasks, right? Or maybe you can, but the reporting is weird? I can't remember. Do you need to introduce a new `@testset` within this task?

NHDaly commented on test/threads_exec.jl:
```diff
@@ -1312,4 +1313,80 @@ end
+    end
+end
+
 end # main testset
```
Maybe you could add a bit more testing around edge cases, like if you create a Task but don't run it, what its numbers look like, or if you create a task and try to run it, but you provide a bad value for the function so it never actually runs, what happens to the numbers.

And what happens if a task throws an exception, do its numbers still get cleaned up correctly (they don't keep incrementing), etc.

nickrobinson251 replied:
note to self: still need to add these tests

nickrobinson251 replied:
okay, added more tests -- please take a look (more suggestions welcome!)

NHDaly commented on test/threads_exec.jl:
```diff
@@ -1312,4 +1313,80 @@ end
+            end_time = time_ns()
+            wall_time_delta = end_time - start_time
+            @test t.is_timing_enabled
+            @test Base.task_cpu_time_ns(t) > 0
+            @test Base.task_wall_time_ns(t) > 0
+            @test Base.task_wall_time_ns(t) >= Base.task_cpu_time_ns(t)
+            @test wall_time_delta > Base.task_wall_time_ns(t)
```
Can you add a tiny test that these aren't still ticking up? That they're fixed to a value once the task ends.

nickrobinson251 replied:
added

NHDaly commented on base/task.jl:
```diff
@@ -1176,3 +1206,20 @@ if Sys.iswindows()
+    if t.is_timing_enabled
+        @assert t.last_dequeued_at != 0
+        t.cpu_time_ns += stopped_at - t.last_dequeued_at
+        t.last_dequeued_at = 0
```
Do you ever check/rely anywhere that this is reset to 0?

It seems unnecessary to me (and we want to keep task-switches as cheap as we can, even though this will probably be very cheap since we'll already have the task resident in-cache, but still writes to main memory are slowish). And maybe someone would even want to look at the last-ran-time even for a sleeping task if they were trying to debug things? Not sure though.

nickrobinson251 replied:
we now use it in `Base.task_cpu_time_ns` to determine if the task is running, so that can report the CPU time of running tasks (without it being out of date til the task yields or finishes)

We could have split added a `task_state_running` for this (to distinguish running from `task_state_runnable`) but that felt like a more intrusive change. We could have added a `is_running::Bool` field for it, but it seemed simpler to me to re-use this field.

nickrobinson251 replied:
update: ignore what i said. removed these after all

NHDaly commented on base/task.jl:
```diff
@@ -972,7 +978,13 @@ function enq_work(t::Task)
 
-schedule(t::Task) = enq_work(t)
+function schedule(t::Task)
+    # user_time -task-(re)scheduled-> wait_time
+    if t.is_timing_enabled && t.first_enqueued_at == 0
+        t.first_enqueued_at = time_ns()
+    end
```
Thanks for having these edge labels, they're very helpful. ❤️ 

If in a future PR we add a distinction between wait-reasons, we would probably put another update here to take us from e.g. waiting on a lock to waiting in the run queue.

🤔 ... which actually makes me think: I think this isn't actually taking the Task `t` from user -> wait_time. The task `t` wasn't running before this function was called.

This function is moving the task from one kind of not-running to a new kind of not-running. In this PR, it's moving it from state CREATED to state WAITING, which we didn't originally include in our state diagram, but i guess we should have.

```suggestion
    # created -task-scheduled-> wait_time
    if t.is_timing_enabled && t.first_enqueued_at == 0
        t.first_enqueued_at = time_ns()
    end
```
And i think in the future, we'd add another possible edge for this:
```
    # waiting on lock -task-scheduled-> waiting in run queue
```
when we add those finer-grained information.

nickrobinson251 replied:
yeah, this should be two arrows, as it can be either being put in the run queue for the first time or because it's being reinserted after e.g. being in `wait(condition)`.

so maybe
```
# created -task-scheduled-> wait_time, or
# blocked_on_conditon -task-scheduled-> wait_time
```
?

NHDaly replied:
Yeah, exactly.

But i would not add the second arrow yet, since we aren't actually recording anything about that edge, right? Your if-condition is only checking for the first arrow.

And in particular, the states on either side of the arrow, "blocked_on_condition" and "in the run queue" are _both_ "wait_time", so i think that we would only want to add an arrow here once we have broken this down into more detailed states.

NHDaly commented on base/task.jl:
```diff
@@ -1018,6 +1030,10 @@ true
 function schedule(t::Task, @nospecialize(arg); error=false)
     # schedule a task to be (re)started with the given value or exception
     t._state === task_state_runnable || Base.error("schedule: Task not runnable")
+    # user_time -task-(re)scheduled-> wait_time
+    if t.is_timing_enabled && t.first_enqueued_at == 0
+        t.first_enqueued_at = time_ns()
+    end
```
Same here i think
```suggestion
    # created -task-scheduled-> wait_time
    if t.is_timing_enabled && t.first_enqueued_at == 0
        t.first_enqueued_at = time_ns()
    end
```

NHDaly commented on base/task.jl:
```diff
@@ -1039,6 +1055,7 @@ tasks.
 """
 function yield()
     ct = current_task()
+    record_cpu_time!(ct)
```
```suggestion
    user_time -yield-> wait_time
    record_cpu_time!(ct)
```
? is it this?

NHDaly commented on base/task.jl:
```diff
@@ -1060,6 +1077,7 @@ Throws a `ConcurrencyViolationError` if `t` is the currently running task.
 """
 function yield(t::Task, @nospecialize(x=nothing))
     current = current_task()
+    record_cpu_time!(current)
```
```suggestion
    # user_time -yield-> wait_time
    record_cpu_time!(current)
```
same here?

NHDaly commented on base/task.jl:
```diff
@@ -1084,6 +1102,9 @@ function yieldto(t::Task, @nospecialize(x=nothing))
     elseif t._state === task_state_failed
         throw(t.result)
     end
+    if t.is_timing_enabled && t.first_enqueued_at == 0
+        t.first_enqueued_at = time_ns()
+    end
```
and here

NHDaly commented on base/task.jl:
```diff
@@ -1097,6 +1118,12 @@ function try_yieldto(undo)
+    # scheduler -task-started-> user
+    # scheduler -task-resumed-> user
+    if ct.is_timing_enabled
+        # @assert ct.last_dequeued_at == 0
```
yeah, per my other comment, i think we can delete this, so that we don't need to waste the perf on zeroing it?

NHDaly commented on base/task.jl:
```diff
@@ -1110,6 +1137,9 @@ end
 
 # yield to a task, throwing an exception in it
 function throwto(t::Task, @nospecialize exc)
+    if t.is_timing_enabled && t.first_enqueued_at == 0
+        t.first_enqueued_at = time_ns()
+    end
```
💡 wow good hunting

NHDaly commented on src/julia.h:
```diff
@@ -2289,6 +2289,16 @@ typedef struct _jl_task_t {
     uint16_t priority;
+    // flag indicating whether or not this to measure the wall and cpu time of this task
+    uint8_t is_timing_enabled;
+    // timestamp this task first entered the run queue (TODO: int32 of ms instead?)
+    uint64_t first_enqueued_at;
```
Do you think it's valuable to also have created_at?

It _might_ be interesting in some cases to see the difference between the created-at and first-scheduled-at timestamps? Maybe someone would be surprised to see that gap being large? Not sure though.

nickrobinson251 replied:
yeah, maybe, but i didn't particularly want to add things i wasn't adding a public API for... what do you think? worth adding? if so, worth adding a public API?

NHDaly replied:
+1 that it should either have a public API or not exist at all.

🤔 My instinct was to say "just add it," but i know that we want to keep the struct smallish if possible for data-locality so I think you're right to leave it off unless we have a use for it.

nickrobinson251 replied:
i've not added it in this PR, but seems trivial to add in a follow up if we decide it'd be worth having (we can add a field that gets set in `jl_new_task`)

NHDaly reviewed (approved):
Coooooooool! 😁 Looks great from my perspective!

I'll let everyone else finish their reviews and your responses and such, but this LGTM from my perspective. 😎

NHDaly commented on src/task.c:
```diff
@@ -1596,6 +1607,17 @@ jl_task_t *jl_init_root_task(jl_ptls_t ptls, void *stack_lo, void *stack_hi)
     ct->reentrant_timing = 0;
+    ct->cpu_time_ns = 0;
+    ct->wall_time_ns = 0;
+    if (ct->is_timing_enabled) {
+        uint64_t now = jl_hrtime();
+        ct->first_enqueued_at = now;
+        ct->last_dequeued_at = now;
+    }
+    else {
+        ct->first_enqueued_at = 0;
+        ct->last_dequeued_at = 0;
+    }
```
Should these also have 
```
# created -task-started-> user_time
```
edges?

Or do the root tasks go to wait_time or straight to user_time? I guess based on setting `last_dequeued_at` they go straight to user_time?

NHDaly commented on base/task.jl:
At first I was going to suggest that it might actually be nice to have an ascii-art state diagram for the state transitions you've added here...

... But I guess for now there'd only be two states: user_time and wait_time, with all the like 15 different arrows you've added.
... so maybe that's actually not super useful right now and it makes more sense to just grep the code base for `-> `.

Which actually, maybe you can just make them more searchable, by adding a keyword? Like:
```
# [Task timings]: user_time -task-finished-> wait_time
```
at the front of each arrow or something?

nickrobinson251 replied:
added something like this -- would love you to review them again 🙏

nickrobinson251 committed:
Enable task timing for root task

nickrobinson251 committed:
Prevent `task_timing(false)` decrementing below zero

nickrobinson251 committed:
Change task-timing flag to require yes/no arg

nickrobinson251 committed:
Rename `timings`->`metrics`, `dequeued`->`started_running`

nickrobinson251 committed:
Update task-state-transition comments

nickrobinson251 committed:
Remove unused debug function

nickrobinson251 committed:
Add NEWS

nickrobinson251 committed:
Add fallback recording of cpu time in `wait()`

nickrobinson251 committed:
Instrument async io

nickrobinson251 committed:
Record task cpu time in `wait()`

nickrobinson251 commented:
Thanks for all the great reviews so far! on wednesday i'll try to get this PR finished up for a final review

nickrobinson251 committed:
fixup! Change task-timing flag to require yes/no arg

nickrobinson251 committed:
fixup! Record task cpu time in `wait()`

nickrobinson251 committed:
Make task metrics flag const and counters atomic

nickrobinson251 commented:
Okay, i think the implementation is here. I need to add more tests, and there are still a few API decisions (and potentially the new APIs should be marked experimental).

nickrobinson251 committed:
Mark the new APIs experimental

nickrobinson251 committed:
Return task metrics as Int or nothing if disabled

nickrobinson251 committed:
Test metrics updating as expected

nickrobinson251 commented on src/task.c:
```diff
@@ -313,6 +313,10 @@ void JL_NORETURN jl_finish_task(jl_task_t *ct)
     JL_PROBE_RT_FINISH_TASK(ct);
     JL_SIGATOMIC_BEGIN();
+    if (ct->metrics_enabled) {
+        assert(ct->first_enqueued_at != 0);
+        ct->finished_at = jl_hrtime();
```
```suggestion
        assert(jl_atomic_load_relaxed(ct->first_enqueued_at) != 0);
        jl_atomic_store_relaxed(&ct->finished_at, jl_hrtime());
```

kpamnany replied:
Think you need an `&` on the relaxed load (in the `assert`).

nickrobinson251 committed:
fixup! Make task metrics flag const and counters atomic

nickrobinson251 committed:
More tests

nickrobinson251 committed:
Update test/core.jl

nickrobinson251 commented on test/threads_exec.jl:
```diff
@@ -1312,4 +1313,231 @@ end
+                    @test cpu_time_i > 0
+                    @test wall_time_i > 0
+                    # task cpu-time cannot be greater than its wall-time
+                    @test wall_time_i >= cpu_time_i
```
😞 
```
  Expression: wall_time_i >= cpu_time_i
   Evaluated: 13247163681 >= 30917448428008467
```
on CI https://buildkite.com/julialang/julia-master/builds/41766#0192e2c2-ae14-4f9f-93dd-9ea96e39642d
```
threads                                           (1) |        started at 2024-10-31T16:21:10.844 on pid 646
Test Summary:                               |  Pass  Total   Time
threads_exec.jl with JULIA_NUM_THREADS == 1 | 40471  40471  47.4s
Test Summary:                               |  Pass  Total   Time
threads_exec.jl with JULIA_NUM_THREADS == 2 | 49137  49137  51.2s
Test Summary:                               |  Pass  Total   Time
threads_exec.jl with JULIA_NUM_THREADS == 4 | 50145  50145  55.6s
Test Failed at /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1521
  Expression: wall_time_i >= cpu_time_i
   Evaluated: 13247163681 >= 30917448428008467
task time counters: lots of spawns: Error During Test at /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1495
  Got exception outside of a @test
  TaskFailedException
      nested task error: There was an error during testing
  Stacktrace:
    [1] sync_end(c::Channel{Any})
      @ Base ./task.jl:601
    [2] macro expansion
      @ ./task.jl:634 [inlined]
    [3] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1507 [inlined]
    [4] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/stdlib/v1.12/Test/src/Test.jl:1707 [inlined]
    [5] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:1496 [inlined]
    [6] macro expansion
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/stdlib/v1.12/Test/src/Test.jl:1707 [inlined]
    [7] top-level scope
      @ /cache/build/tester-amdci4-11/julialang/julia-master/julia-ea5e5e9e47/share/julia/test/threads_exec.jl:33
    [8] include(mod::Module, _path::String)
      @ Base ./Base.jl:590
    [9] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:329
   [10] _start()
      @ Base ./client.jl:558
```

nickrobinson251 replied:
And another 
```
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-26ec30e9b7/share/julia/test/threads_exec.jl:1514
  Expression: cpu_time_i > 0
   Evaluated: 0 > 0
task time counters: lots of spawns: Error During Test at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-26ec30e9b7/share/julia/test/threads_exec.jl:1491
  Got exception outside of a @test
  TaskFailedException
      nested task error: There was an error during testing
  Stacktrace:
    [1] sync_end(c::Channel{Any})
      @ Base ./task.jl:601
    [2] macro expansion
      @ ./task.jl:634 [inlined]
    [3] macro expansion
      @ /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-26ec30e9b7/share/julia/test/threads_exec.jl:1503 [inlined]
```

nickrobinson251 replied:
```
  Expression: cpu_time_i > 0
   Evaluated: 0 > 0
```

Hoping this one at least was because we were setting the `t._state` to done before the final update to `t.cpu_time_ns`, so there was a race where `wait(t)` could return (because `wait(t)` is `while !istaskdone(t)`) before we've set `t.cpu_time_ns` to a non-zero value. Moved the update to be in finish_task before we set the `_state`

nickrobinson251 committed:
Introduce `current_task_*_time_ns` for non-racy continuous metrics

for the running task only.
Other tasks report metrics that are non-continuous
but updated only on task switches.

nickrobinson251 committed:
Consolidate names again

nickrobinson251 committed:
Update CPU time on finish before setting task done

nickrobinson251 committed:
Remove TODO comment

nickrobinson251 committed:
Add new functions to docs

nickrobinson251 committed:
fixup! Add new functions to docs

kpamnany committed:
Return `UInt64` instead of `Int`

kpamnany committed:
Always return a `UInt64` type (even when returning 0)

NHDaly committed:
Convert task timestamps to UInt32 delta-times since system boot.

vchuravy commented on base/task.jl:
```diff
@@ -1181,3 +1205,20 @@ if Sys.iswindows()
+# update the `cpu_time_ns` field of `t` to include the time since it last started running.
+function record_cpu_time!(t::Task)
+    if t.metrics_enabled && !istaskdone(t)
+        @atomic :monotonic t.cpu_time_ns += delta_time_ns() - t.last_started_running_at
```
Do we have to share the implementation between C and Julia?

Can we just export a single global and access it from Julia?

NHDaly replied:
Good idea, that's probably faster.
(So duplicate the `delta_time_ns()` implementation, you mean?)

We will have to ccall `jl_hrtime()` either way though, so i'm not sure if it will actually save us anything?

vchuravy replied:
Yeah, but that is one function call, instead of one and then another. Since `jl_hrtime` won't be inlined.

NHDaly replied:
💡 yep, thanks, makes sense! 👍 I'll do that then.

NHDaly replied:
moot, because this change is reverted

vchuravy commented on base/task.jl:
```diff
@@ -1181,3 +1205,20 @@ if Sys.iswindows()
+        @atomic :monotonic t.cpu_time_ns += delta_time_ns() - t.last_started_running_at
+    end
+    return t
+end
```
```suggestion
end

```

vchuravy commented on src/jltypes.c:
```diff
@@ -3762,8 +3762,13 @@ void jl_init_types(void) JL_GC_DISABLED
-                                        "priority"),
-                        jl_svec(16,
+                                        "priority",
+                                        "metrics_enabled",
+                                        "first_enqueued_at",
+                                        "last_started_running_at",
+                                        "cpu_time_ns",
+                                        "finished_at"),
```
Would it make sense to instead store this state out-of-line and just have a single pointer here?

You could be cute and have a Union{Nothing,Metric} for the metric enabled query.
Alternativly these values could live in task-local-storage.

NHDaly replied:
We originally proposed that, yeah, but they were moved inline. I don't remember why. Maybe to dodge the extra cache-miss on the task switches?

CC @vtjnash

NHDaly committed:
Update base/task.jl

Co-authored-by: Valentin Churavy <vchuravy@users.noreply.github.com>

vtjnash commented on src/julia.h:
```diff
@@ -2274,6 +2274,15 @@ typedef struct _jl_handler_t {
 
 #define JL_RNG_SIZE 5 // xoshiro 4 + splitmix 1
 
+// delta_timestamp values are nanoseconds since the boot-time of the julia runtime system.
```
That's an overflow every 4 seconds (vs uint64_t which was an overflow every 5 centuries). I think you need to reduce precision somewhat?

NHDaly replied:
Yeah, i realized the same. I think uint32 is going to be just too small to be useful for measuring time. Discussing [on slack](https://julialang.slack.com/archives/C02H0E8S2NA/p1732133591721809?thread_ts=1732132591.394879&cid=C02H0E8S2NA).

NHDaly replied:
reverted

NHDaly committed:
Revert "Convert task timestamps to UInt32 delta-times since system boot."

This reverts commit d23a96f481b263a72a51c930bc1c7daa6bb68a8b.

NHDaly committed:
Try adding `alignas(8)` to force correct alignment and work around GCC bug

topolarity commented on base/experimental.jl:
```diff
@@ -494,4 +494,78 @@ function entrypoint(@nospecialize(argt::Type))
+!!! compat "Julia 1.12"
+    This method was added in Julia 1.12.
+"""
+function task_cpu_time_ns(t::Task=current_task())
```
This name seems like it might be a bit mis-leading - What about `task_running_time_ns()`?

As the note clarifies, this is the time that the task has been "in the running state" (i.e. scheduled by Julia) but not the time the task has actually spent running on-cpu (i.e. scheduled by the OS)

It's plausible to me that one day we might want to hook up `perf` with our scheduler so that you can get the actual instruction / branch / tsc spent on-core, so it's probably good to leave room for the future names

NHDaly replied:
That seems like good feedback to me. 👍 
I'm supportive of a name change.

I like `task_running_time_ns()`. I also like `task_executing_time_ns()` or maybe `task_scheduled_time_ns()`?
I think `task_running_time_ns` is the best I've seen so far.

kpamnany replied:
Changed.

kpamnany committed:
Revert "Try adding `alignas(8)` to force correct alignment and work around GCC bug"

This reverts commit 4a5495022d4f64726ddaa66590859dfc72bc22f7.

NHDaly commented on src/jlapi.c:
```diff
@@ -809,6 +809,28 @@ JL_DLLEXPORT uint64_t jl_cumulative_recompile_time_ns(void)
+JL_DLLEXPORT void jl_task_metrics_enable(void)
+{
+    // Increment the flag to allow reentrant callers.
+    jl_atomic_fetch_add(&jl_task_metrics_enabled, 1);
```
We're going to try this approach, it seems like the best path forward. 👍 
@kpamnany is going to push up some commits here. Thanks valentin!

NHDaly replied:
(jk, we didn't do this because it was too big of a change. I guess there are bootstrapping challenges relating to adding fields to the Task struct. But we'll probably make this change the next time we add more metrics)

kpamnany committed:
Reorder task structure and add padding

To ensure that the task metrics do not cross a cache line boundary.
Also rename `cpu_time_ns` to `running_time_ns`.

kpamnany committed:
More changes of `task_cpu_time_ns` to `task_running_time_ns`

kpamnany committed:
Move calls to `record_running_time!`

To prevent miscounting in case of exceptions/errors in task switching.

kpamnany commented:
I've addressed the failure on i686 (see https://github.com/JuliaLang/julia/issues/56626) by following @Keno's [suggestion](https://github.com/JuliaLang/julia/issues/56626#issuecomment-2489805684) to align the outer struct, which is `jl_task_t` in this case. I reordered a few fields to be on word boundaries and inserted explicit padding bytes.

Please take a final look @vtjnash and @topolarity? In particular, [changes to `jl_task_type`](https://github.com/JuliaLang/julia/pull/56320/commits/788f46fe9599577320eaba435d4e9dbf4f5c2ae3#diff-882927c6e612596e22406ae0d06adcee88a9ec05e8b61ad81b48942e2cb266e9R3749) and [changes to `jl_task_t`](https://github.com/JuliaLang/julia/pull/56320/commits/788f46fe9599577320eaba435d4e9dbf4f5c2ae3#diff-42a3fc6ebd78315e40602ab4108f5cd380ada14ff81987b696d7b4ee9372a4deL2279).

kpamnany commented:
CI failures seem unrelated to this PR.

JeffBezanson commented:
I wouldn't expect this to have a lot of overhead but it would be nice to have a simple benchmark. Of course there might be a bit of overhead when timing is turned on, but it would be interesting to see how much.

kpamnany commented:
> I wouldn't expect this to have a lot of overhead but it would be nice to have a simple benchmark. Of course there might be a bit of overhead when timing is turned on, but it would be interesting to see how much.

The overhead is primarily in task switching, where the metrics are updated. Good ol' `pfib` suffers the most.

I don't see any measurable difference in performance when task metrics are off when compared to `master`.

Besides `pfib`, here are some of the benchmarks from the (not yet complete) [MultithreadingBenchmarks.jl](https://github.com/kpamnany/MultithreadingBenchmarks.jl), using 6 threads.

Benchmark | Metrics | No metrics
--- | --- | ---
`pfib.jl` -- fib(12) | 123.1 μs | 69.7 μs
`psort.jl` | 288.292 ms | 286.990 ms
`sieve.jl` | 393.7 ms | 267.124 ms
`simple_independent.jl` | 30.690 ms | 29.978 ms

nickrobinson251 commented:
So great to see this landed! 🚀  Thanks for all the great reviews and help, and big thank you to @kpamnany for taking this over when i needed to go on leave ❤️
